scalableminds/webknossos
# Time-track movements in the 3D viewport

- Added a new `updateTdCamera` update action which is triggered if the user modified the 3D view, so that these actions can be time tracked. The action is empty otherwise.
- While I was at it I spent some love on the version view, added a lot of missing descriptions for certain update actions (tree/group visibility, user bounding boxes, fallback layer removal, node updates, edge deletion/addition) and extended the tree deletion message as by now multiple trees can be deleted at once.
- I added save queue compaction for `updateNode` update actions as otherwise lots of new versions are created if the radius of a node is changed or a node is moved. `updateTdCamera` actions are also compacted and only the last one within one save interval is kept.
- I've also noticed that the Undo of a tree deletion is misinterpreted as an NML upload by the version view (as multiple nodes are created at once which is not possible otherwise). Looking at the update actions one cannot be distinguished from the other, so I'm not sure how to fix this.

TODO:
- [x] Rotation in the 3D view is not picked up as a change yet, because the action that is used for that is also used when hovering over the 3D view and when initializing the tracing (those should not be counted as actions). I'll probably dispatch another action for the 3D view rotation so that it can be recognized and told apart from the other actions. _(Solved by introducing a duplicate action that doesn't trigger the save queue diffing which is used accordingly)_

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Move/Zoom/Rotate the 3D view - the save button should indicate that there are changes which need to be saved.
- Create a task and exclusively do stuff in the 3D view - time should be tracked.
- Check the version view to see whether the action is correctly labeled and also that subsequent movements in the 3D view within one save interval are compacted to only one version.
- Do some of the other mentioned actions with new descriptions/compactions and check the version view afterwards.

### Issues:
- fixes #4825 

------
